ribs/archives/_dns_archive.py

```diff
@@ -0,0 +1,587 @@
+        )
+
+        # Set up statistics -- objective_sum is the sum of all objective values in the
+        # archive; it is useful for computing qd_score and obj_mean.
+        self._best_elite = None
+        self._objective_sum = None
+        self._stats_reset()
+
+    ## Properties inherited from ArchiveBase ##
+
+    @property
+    def field_list(self) -> list[str]:
+        return self._store.field_list_with_index
+
+    @property
+    def dtypes(self) -> dict[str, np.dtype]:
+        return self._store.dtypes_with_index
+
+    @property
+    def stats(self) -> ArchiveStats:
+        return self._stats
+
+    @property
+    def empty(self) -> bool:
+        return len(self._store) == 0
+
+    ## Properties that are not in ArchiveBase ##
+    ## Roughly ordered by the parameter list in the constructor. ##
+
+    @property
+    def best_elite(self) -> SingleData | None:
+        """The elite with the highest objective in the archive.
+
+        None if there are no elites in the archive.
+        """
+        return self._best_elite
+
+    @property
+    def k_neighbors(self) -> int:
+        """The number of fitter neighbors for computing DNS."""
+        return self._k_neighbors
+
+    @property
+    def space(self) -> str:
+        """Space used for distance calculations ("measures" or "solution")."""
+        return self._space
+
+    @property
+    def capacity(self) -> int:
+        """Fixed number of solutions stored in this archive."""
+        return self._store.capacity
+
+    @property
+    def cells(self) -> int:
+        """Total capacity of the archive (for coverage/statistics)."""
+        return self.capacity
+
+    @property
+    def qd_score_offset(self) -> float:
+        """Subtracted from objective values when computing the QD score."""
+        return self._qd_score_offset
+
+    ## dunder methods ##
+
+    def __len__(self) -> int:
+        return len(self._store)
+
+    def __iter__(self) -> Iterator[SingleData]:
+        return iter(self._store)
+
+    ## Utilities ##
+
+    def _stats_reset(self) -> None:
+        """Resets the archive stats."""
+        self._best_elite = None
+        self._objective_sum = np.asarray(0.0, dtype=self.dtypes["objective"])
+        self._stats = ArchiveStats(
+            num_elites=0,
+            coverage=np.asarray(0.0, dtype=self.dtypes["objective"]),
+            qd_score=np.asarray(0.0, dtype=self.dtypes["objective"]),
+            norm_qd_score=np.asarray(0.0, dtype=self.dtypes["objective"]),
+            obj_max=None,
+            obj_mean=None,
+        )
+
+    def _stats_update(self, new_objective_sum: Float, new_best_index: Float) -> None:
+        """Updates statistics.
+
+        Update is based on a new sum of objective values (new_objective_sum) and the
+        index of a potential new best elite (new_best_index).
+        """
+        _, new_best_elite_raw = self._store.retrieve([new_best_index])
+        new_best_elite_dict = cast(BatchData, new_best_elite_raw)
```

Comment:
Any reason for casting here? I think you can just revert it to match what is done in the other archives.